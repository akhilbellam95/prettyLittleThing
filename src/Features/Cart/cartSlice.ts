import { createSlice } from '@reduxjs/toolkit';

interface CartItem {
  id: number;
  quantity: number;
};

export interface CartState {
  cartItems: CartItem[],
  total: number
}

export const cartInitialState: CartState = {
  cartItems: [] as CartItem[],
  total: 0,
};

export const cartSlice = createSlice({
  name: 'cart',
  initialState: cartInitialState,
  reducers: {
    addToCart: (state, action) => {
      const { id } = action.payload;
      const existingItem = state.cartItems.find((item) => item.id === id);
      let total = state.total;
      if (existingItem) {
        existingItem.quantity++;
      } else {
        state.cartItems.push({ id, quantity: 1 });
      }
      total += action.payload.price;
      state.total = parseFloat(total.toFixed(2));
    },
    removeFromCart: (state, action) => {
      const itemIndex = state.cartItems.findIndex((el) => el.id === action.payload.id);
      if (itemIndex !== -1) {
        const item = state.cartItems[itemIndex];
        let total = state.total;
        if (item.quantity === 1) {
          state.cartItems.splice(itemIndex, 1);
        } else {
          item.quantity--;
        }
        total -= action.payload.price;
        state.total = parseFloat(total.toFixed(2));
      }
    },
    updateTotal: (state, action) => {
      state.total = action.payload;
    },
  },
});

export const { addToCart, removeFromCart, updateTotal } = cartSlice.actions;

export default cartSlice.reducer;
