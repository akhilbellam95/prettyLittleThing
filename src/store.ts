// eslint-disable-next-line prettier/prettier
import { configureStore } from '@reduxjs/toolkit';
import cartReducer from './Features/Cart/cartSlice';
import productReducer from './Features/Product/productSlice';

export const store = configureStore({
  reducer: {
    cart: cartReducer,
    product: productReducer,
  },
  devTools: process.env.NODE_ENV !== 'production',
});

export type AppDispatch = typeof store.dispatch;

export type RootState = ReturnType<typeof store.getState>;
