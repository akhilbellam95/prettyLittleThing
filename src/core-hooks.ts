import { configureStore, createStore } from "@reduxjs/toolkit";
import { useSelector } from "react-redux";
import { TypedUseSelectorHook, useDispatch } from "react-redux";
import { CartState } from "./Features/Cart/cartSlice";
import { ProductsState } from "./Features/Product/productSlice";

export interface ApplicationState {
  cart: CartState;
  product: ProductsState;
}

export const useAppDispatch = () =>
  useDispatch<ReturnType<typeof configureStore>['dispatch']>();

export const useAppSelector: TypedUseSelectorHook<ApplicationState> =
  useSelector;