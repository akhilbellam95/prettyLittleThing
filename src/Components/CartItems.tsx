import React from 'react';
import { FlatList, StyleSheet, Text, View } from 'react-native';
import { updateTotal } from '../Features/Cart/cartSlice';

import { ApplicationState, useAppDispatch, useAppSelector } from '../core-hooks';
import { CartItem } from './CartItem';

export type ItemProps = {
  cartItem: {
    id: number;
    quantity: number;
  };
};

const CartItems = () => {
  const { cartItems, total } = useAppSelector((state: ApplicationState) => state.cart);

  const dispatch = useAppDispatch();

  if (cartItems?.length === 0 || !cartItems) dispatch(updateTotal(0));

  return (
    <View style={styles.container}>
      <View style={{ flex: 6 }}>
        <View style={{ height: 30 }} />
        <Text style={styles.cartItems}>{`Cart Items (${cartItems?.length})`}</Text>
        <FlatList data={cartItems} renderItem={({ item }) => <CartItem cartItem={item} />} />
      </View>

      <View style={styles.cartDetails}>
        <Text style={styles.cartTotal}>Cart Total</Text>
        <Text>{total}</Text>
      </View>
    </View>
  );
};

export default CartItems;

export const styles = StyleSheet.create({
  cartDetails: {
    flex: 1,
    flexDirection: 'row',
    marginTop: 20,
  },
  cartItem: {
    borderWidth: 2,
    borderColor: 'black',
    paddingHorizontal: 10,
    backgroundColor: 'white',
    marginBottom: 10,
    flexDirection: 'row',
    paddingVertical: 10,
  },
  cartItems: {
    fontWeight: 'bold',
    marginBottom: 20,
  },
  cartTotal: {
    width: 150,
    textAlign: 'center',
  },
  container: {
    flex: 1,
    width: '90%',
  },
});
