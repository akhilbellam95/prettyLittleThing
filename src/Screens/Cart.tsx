import { StyleSheet, Text, View } from 'react-native';
import React from 'react';
import CartItems from '../Components/CartItems';
import { ApplicationState, useAppSelector } from '../core-hooks';

const Cart = () => {
  const { cartItems } = useAppSelector((state: ApplicationState) => state.cart);

  return (
    <View style={styles.container}>
      {cartItems.length > 0 ? <CartItems /> : <Text>Cart is empty</Text>}
    </View>
  );
};

export default Cart;

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
  },
});
