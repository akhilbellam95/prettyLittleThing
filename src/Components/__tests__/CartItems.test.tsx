import { render } from '@testing-library/react-native';
import React from 'react';
import { Provider } from 'react-redux';
import { store } from '../../store';
import CartItems from '../CartItems';
import { useAppDispatch, useAppSelector } from '../../core-hooks';
import { mockProducts } from '../../mocks/data';
import { Text } from 'react-native';
import { updateTotal } from '../../Features/Cart/cartSlice';

jest.mock('../../core-hooks', () => ({
  useAppSelector: jest.fn(),
  useAppDispatch: jest.fn(),
}));

const dispatch = jest.fn();

describe('CartItems component testing', () => {
  const mockDispatch = jest.fn();
  const mockSelector = jest.fn();

  beforeAll(() => {
    (useAppSelector as jest.Mock).mockImplementation(mockSelector);
    (useAppDispatch as jest.Mock).mockReturnValue(mockDispatch);
  });

  it('renders correctly', () => {
    mockSelector.mockReturnValue({
      cartItems: [
        { id: 1, quantity: 2 },
        { id: 2, quantity: 1 },
      ],
      total: 30,
    });

    const snap: unknown = render(
      <Provider store={store}>
        <CartItems />
      </Provider>,
    ).toJSON();
    expect(snap).toMatchSnapshot();
  });

  it('renders correctly when cart is empty', () => {
    mockSelector.mockReturnValue({
      cartItems: [],
      total: 0,
    });

    const snap: unknown = render(
      <Provider store={store}>
        <Text>Cart is Empty</Text>
      </Provider>,
    ).toJSON();
    expect(snap).toMatchSnapshot();
  });

  it('dispatch total as 0 when there are no items in cart', () => {
    mockSelector.mockReturnValue({
      cartItems: [],
    });
    store.dispatch(updateTotal(0));
    const state = store.getState()['cart'];
    expect(state.total).toBeGreaterThanOrEqual(0);
  });
});
