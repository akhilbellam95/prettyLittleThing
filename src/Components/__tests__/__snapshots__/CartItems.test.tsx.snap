// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartItems component testing renders correctly 1`] = `
<View
  style={
    {
      "flex": 1,
      "width": "90%",
    }
  }
>
  <View
    style={
      {
        "flex": 6,
      }
    }
  >
    <View
      style={
        {
          "height": 30,
        }
      }
    />
    <Text
      style={
        {
          "fontWeight": "bold",
          "marginBottom": 20,
        }
      }
    >
      Cart Items (2)
    </Text>
    <RCTScrollView
      data={
        [
          {
            "id": 1,
            "quantity": 2,
          },
          {
            "id": 2,
            "quantity": 1,
          },
        ]
      }
      getItem={[Function]}
      getItemCount={[Function]}
      keyExtractor={[Function]}
      onContentSizeChange={[Function]}
      onLayout={[Function]}
      onMomentumScrollBegin={[Function]}
      onMomentumScrollEnd={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      removeClippedSubviews={false}
      renderItem={[Function]}
      scrollEventThrottle={50}
      stickyHeaderIndices={[]}
      viewabilityConfigCallbackPairs={[]}
    >
      <View>
        <View
          onFocusCapture={[Function]}
          onLayout={[Function]}
          style={null}
        >
          <View />
        </View>
        <View
          onFocusCapture={[Function]}
          onLayout={[Function]}
          style={null}
        >
          <View />
        </View>
      </View>
    </RCTScrollView>
  </View>
  <View
    style={
      {
        "flex": 1,
        "flexDirection": "row",
        "marginTop": 20,
      }
    }
  >
    <Text
      style={
        {
          "textAlign": "center",
          "width": 150,
        }
      }
    >
      Cart Total
    </Text>
    <Text>
      30
    </Text>
  </View>
</View>
`;

exports[`CartItems component testing renders correctly when cart is empty 1`] = `
<Text>
  Cart is Empty
</Text>
`;
